pkg/provisioner/ironic/ironic.go

```diff
@@ -1346,3 +1368,110 @@ func (p *ironicProvisioner) PowerOff() (result provisioner.Result, err error) {
 
 	return result, nil
 }
+
+func (p *ironicProvisioner) buildAndValidateCleanSteps(hwConfigs []interface{}) (cleanSteps []nodes.CleanStep) {
+	for _, config := range hwConfigs {
+		switch config.(type) {
+		case metal3v1alpha1.RAIDConfig:
+			raidArgs := config.(metal3v1alpha1.RAIDConfig)
+			// Retrieve RAID configuration in form of Ironic's clean step
+			raidCleanSteps, err := p.generateRAIDCleanSteps(&raidArgs)
+			if err != nil {
+				log.Info("Failed to retrieve RAID configuration", "Error:", err)
+			} else {
+				cleanSteps = append(cleanSteps, raidCleanSteps...)
+			}
+		}
+	}
+	return
+}
+
+// generateRAIDCleanSteps verify the given RAID configuration
+func (p *ironicProvisioner) generateRAIDCleanSteps(raidConfig *metal3v1alpha1.RAIDConfig) (cleanSteps []nodes.CleanStep, err error){
+	if &raidConfig == nil {
+		return nil, nil
+	}
+	logicalDisks := []map[string]interface{}{}
+	if raidConfig.RootVolume != nil {
+		rootDisk, err := p.buildRAIDVolumeArgs(raidConfig.RootVolume, true)
+		if err != nil {
+			return nil, err
+		}
+		logicalDisks = append(logicalDisks, rootDisk)
+	}
+	for _, volume := range raidConfig.Volumes {
+		volume, err := p.buildRAIDVolumeArgs(&volume, false)
+		if err != nil {
+			return nil, err
+		}
+		logicalDisks = append(logicalDisks, volume)
+	}
+	if len(logicalDisks) > 0 {
+		// ‘create_configuration’ doesn’t remove existing disks. It is recommended
+		// to add ‘delete_configuration’ before ‘create_configuration’ to make sure
+		// that only the desired logical disks exist in the system after manual cleaning.
+		cleanSteps = append(
+			cleanSteps,
+			nodes.CleanStep{
+				Interface: "raid",
+				Step:      "delete_configuration",
+				Args:      nil,
+			},
+			nodes.CleanStep{
+				Interface: "raid",
+				Step: "create_configuration",
+				Args: map[string]interface{}{"logical_disks": logicalDisks},
+			},
+		)
+	}
+	return cleanSteps, nil
+}
+
+// A private method to build the Args in CleanStep for RAID configuration
+func (p *ironicProvisioner) buildRAIDVolumeArgs(
```

Comment:
This method could use some unit tests.

Reply:
I got it. I will make some unit tests for these method later

Reply:
Done. I have add unit tests for these method